### PR TITLE
implement back buttons for AddContactPage and MainPage

### PR DESCRIPTION
**Pull request by Trolldemorted**
@golf1052 what do you think?

### CODE REVIEW

File: Signal-Windows/ViewModels/MainPageViewModel.cs [hunk 1]
```diff
@@ -67,6 +69,62 @@ public void AddThread(SignalThread contact)
...
+            }
+        }
+
+        public void SwitchToStyle(MainPageStyle newStyle)
+        {
+            if (newStyle == MainPageStyle.Narrow)
+            {
+                if (SelectedThread != null)
+                {
+                    SystemNavigationManager.GetForCurrentView().AppViewBackButtonVisibility = AppViewBackButtonVisibility.Visible;
```
Comment (golf1052) [inline Signal-Windows/ViewModels/MainPageViewModel.cs]: I would suggest putting this in Utils. Something like:

```
public static void EnableBackButton()
{
    SystemNavigationManager.GetForCurrentView().AppViewBackButtonVisibility = AppViewBackButtonVisibility.Visible;
}
```
```
public static void DisableBackButton()
{
    SystemNavigationManager.GetForCurrentView().AppViewBackButtonVisibility = AppViewBackButtonVisibility.Collapsed;
}
```

```

Comment (golf1052) [inline Signal-Windows/ViewModels/MainPageViewModel.cs]: Utils should also be static since I assume it's only going to contain static things.

File: Signal-Windows/ViewModels/MainPageViewModel.cs [hunk 2]
```diff
@@ -67,6 +69,62 @@ public void AddThread(SignalThread contact)
...
+                SystemNavigationManager.GetForCurrentView().AppViewBackButtonVisibility = AppViewBackButtonVisibility.Collapsed;
+                SystemNavigationManager.GetForCurrentView().BackRequested -= BackButton_Click;
+                ThreadListVisibility = Visibility.Visible;
+                ThreadVisibility = Visibility.Visible;
+            }
+        }
+
+        internal async void BackButton_Click(object sender, BackRequestedEventArgs e)
+        {
+            using (await ActionInProgress.LockAsync())
```
Comment (golf1052) [inline Signal-Windows/ViewModels/MainPageViewModel.cs]: Nothing in here is async, why is this using an AsyncLock? Also why use a lock at all?

File: Signal-Windows/ViewModels/MainPageViewModel.cs [hunk 3]
```diff
@@ -67,6 +69,62 @@ public void AddThread(SignalThread contact)
...
+            {
+                SelectedThread = null;
+                Thread.DisposeCurrentThread();
```
Comment (golf1052) [inline Signal-Windows/ViewModels/MainPageViewModel.cs]: Maybe SignalThread should be renamed to SignalConversation? Using thread everywhere could very easily be confused with Threading.Thread (which happened to me here ☹)

File: Signal-Windows/Views/LinkPage.xaml.cs [hunk 4]
```diff
@@ -26,9 +27,10 @@ public LinkPageViewModel Vm
             }
         }
 
-        protected override void OnNavigatedTo(NavigationEventArgs navEvent)
+        protected override void OnNavigatedTo(NavigationEventArgs e)
         {
-            base.OnNavigatedTo(navEvent);
+            base.OnNavigatedTo(e);
+            SystemNavigationManager.GetForCurrentView().AppViewBackButtonVisibility = AppViewBackButtonVisibility.Collapsed;
```
Comment (golf1052) [inline Signal-Windows/Views/LinkPage.xaml.cs]: We should allow the back button on the link page in case the user wants to go back to the start screen.

File: Signal-Windows/ViewModels/MainPageViewModel.cs [hunk 5]
```diff
@@ -67,6 +69,62 @@ public void AddThread(SignalThread contact)
             ThreadsDictionary[contact.ThreadId] = contact;
         }
 
+        public MainPageStyle GetViewStyle(Size s)
+        {
+            if (s.Width <= 640)
```
Comment (golf1052) [inline Signal-Windows/ViewModels/MainPageViewModel.cs]: Should move this magic number to Utils

File: Signal-Windows/Views/RegisterPage.xaml.cs [hunk 6]
```diff
@@ -25,6 +27,12 @@ public RegisterPageViewModel Vm
             }
         }
 
+        protected override void OnNavigatedTo(NavigationEventArgs e)
+        {
+            base.OnNavigatedTo(e);
+            SystemNavigationManager.GetForCurrentView().AppViewBackButtonVisibility = AppViewBackButtonVisibility.Collapsed;
```
Comment (golf1052) [inline Signal-Windows/Views/RegisterPage.xaml.cs]: Same as with the link page. The user should be able to go back to the start screen.

**Comment by golf1052** [comment]
Finished adding my comments

File: Signal-Windows/Views/LinkPage.xaml.cs (hunk 4 above)
Comment (Trolldemorted) [inline Signal-Windows/Views/LinkPage.xaml.cs]: Yes, but we have to ensure that we have terminated any lingering connection to OWS servers and leave the database in a consistent way before navigating - i will have a look as soon as i find the time.

File: Signal-Windows/Views/RegisterPage.xaml.cs (hunk 6 above)
Comment (Trolldemorted) [inline Signal-Windows/Views/RegisterPage.xaml.cs]: Will do, same conditions as for the linking page apply though

File: Signal-Windows/ViewModels/MainPageViewModel.cs (hunk 3 above)
Comment (Trolldemorted) [inline Signal-Windows/ViewModels/MainPageViewModel.cs]: I think Signal-Android names them this way, but good point - should go into a separate commit, though

File: Signal-Windows/ViewModels/MainPageViewModel.cs (hunk 2 above)
Comment (Trolldemorted) [inline Signal-Windows/ViewModels/MainPageViewModel.cs]: We can indeed drop this lock, but we have to be cautious during the handling of composed and received messages, as without this lock the selected thread might change while the message handling is awaiting something.
This though means that everything executed on the UI task **must not** rely on SelectedThread not changing if it uses await. We might want to devise a better lock scheme or document the one we have properly.

**Comment by Trolldemorted** [comment]
I have amended my commit! What do you think? Does it work on your lumia? How is the responsiveness?

Support for link/registration page, the signal-thread rename and the lock cleanup not yet included, as they should go in separate commits and come with some documentation.

**Comment by golf1052** [comment]
Sorry I didn't respond to this. It looks good. The contact photo should be shrunk in size when on a phone.
